### DISCUSSION TITLE
Build custom android app with my own bundle name

### DISCUSSION THREAD

**Discussion by realbitt** [strongswan/strongswan#938] [General]
Two months working to build my own android application using strongswan, two month ago searching the easiest way to create android app. finally I found flutter package used strongswan 
this link , https://pub.dev/packages/flutter_vpn with simple functions connect, disconnect and vpn_stat WOW 
everything works in development until I ran into a problem while finalizing the application, that application does not work when building in release mode, Unfortunately, I discovered this too late. no luck to found any solution till post.

**some of flutter build log:**
```

PS D:\Development\Flutter\thrustvpn\thrustvpn> flutter run --release
Launching lib\main.dart on Redmi Note 8 Pro in release mode...
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Running Gradle task 'assembleRelease'...                           74.9s
√  Built build\app\outputs\flutter-apk\app-release.apk (9.2MB).
Installing build\app\outputs\flutter-apk\app.apk...                 5.9s
...
**highlighted error :**
...
/strongswan.thrustv(29311): runtime.cc:659] Pending exception java.lang.NoSuchMethodError: no static method "Lorg/strongswan/android/logic/CharonVpnService;.getAndroidVersion()Ljava/lang/String;"
....
400+ lines 
....
F/strongswan.thrustv(29311): runtime.cc:667] JNI DETECTED ERROR IN APPLICATION: mid == null
F/strongswan.thrustv(29311): runtime.cc:667]     in call to CallStaticObjectMethod
F/strongswan.thrustv(29311): runtime.cc:667]     from java.lang.String java.lang.Runtime.nativeLoad(java.lang.String, java.lang.ClassLoader, java.lang.Class)

Application finished.

PS D:\Development\Flutter\thrustvpn\thrustvpn>    
```
**full log file :**
[Full Log](https://www.noonsmartsystems.com/logs/flutter_build_log.txt)

I closed the flutter page and went back to the original source that I succeeded in a while ago to make a native build without any problem and it was run on the my device (debug and release) which is the original version of your Android application that comes with the source (frontends) 
I started my attempts to make my own application (own bundle name) and I did these steps and I have not succeeded so far.

**Second attempt:**
- I changed the name and bundle name of the original application accurately and reached the point that it is build's successfully and installing application on my own device, but when I run the application it comes out suddenly as soon as I run it, the same problem as the previous one that occurred in the flutter and almost the same error.

**Third attempt:**
I made a new, clean app and copied the libraries (*.so) from the original build. Under the jnilibs folder, as I read, these libraries can be used to communicate with the native functions and methods, and I did not reach far from this point. I did not find any sources that explain this clearly.

**Third attempt:**
The last attempt is to convert the original application source into a library with a few modifications such as deleting the applicationId and changing the apply plugin: **'com.android.application'**
**Apply plugin: 'com.android.library '** and then import this library into my new application, so far I'm trying this method and I haven't succeeded in importing it as a library, but still trying.

**Conclusion :**
My attempts in every direction so far have not succeeded. I just want a solution or advice or any hint. What is the correct way that I can reach my goal without wasting more effort and time.

thank you

**Comment by realbitt** [comment]
i solved the problem, thanks

**Comment by tobiasbrunner** [comment]
That package has nothing to do with the strongSwan project so if you have trouble, please contact its developers directly.

Also note that this does not change strongSwan's license.  strongSwan is still GPLv2 licensed even if you don't directly use its source code (the package seems to load pre-built libraries).  So without commercial license, you still have to provide the complete source code of your app that's based on it to your users.
